ecosystem/purescript-0.14-migration-guide.md

```diff
@@ -0,0 +1,618 @@
+
+```purs
+foreign import kind Boolean
+foreign import data True :: Boolean
+foreign import data False :: Boolean
+```
+
+and should be rewritten as an empty data declaration:
+
+```purs
+data Boolean
+foreign import data True :: Boolean
+foreign import data False :: Boolean
+```
+
+The old unary `# Type` syntax for row kinds is deprecated, and it should instead be written as `Row Type`.
+
+```diff
+- class ListToRow (list :: RowList) (row :: # Type)
++ class ListToRow (list :: RowList) (row :: Row Type)
+```
+
+Finally, all type-level declarations are now generalized.
+
+```purs
+data Proxy a
+```
+
+This previously had the `Type`-defaulted kind `Type -> Type`. Now this will be generalized to `forall k. k -> Type`. It is better to be explicit about polymorphism by writing signatures, and the compiler will warn about missing kind signatures when it infers polymorphic kinds.
+
+To avoid the compiler warning, provide a kind signature:
+
+```purs
+data Proxy :: forall k. k -> Type
+data Proxy a = Proxy
+```
+
+Classes can have signatures too, and they must end with a `Constraint` kind instead of `Type`. For example, here's the new definition of `Prim.Row.Cons`:
+
+```purs
+class Cons :: forall k. Symbol -> k -> Row k -> Row k -> Constraint
+class Cons label a tail row | label a tail -> row, label row -> a tail
+```
+
+### Coercible & Role Annotations
+
+PureScript 0.14 introduces a new compiler-solved class, `Prim.Coerce.Coercible`, to relate types with the same runtime representation. You can now use `Safe.Coerce.coerce` from the new `safe-coerce` library in place of `Unsafe.Coerce.unsafeCoerce` to turn an `a` into a `b` when `Coercible a b` holds.
+
+If you are currently using `unsafeCoerce` to transform two types with the same runtime representation, you should consider replacing it with `coerce`.
+
+The `Coercible` implementation introduces role annotations for type parameters (to learn more about roles, see the compiler release notes). Role annotations can loosen or strengthen the compiler-inferred roles of type parameters.
+
+For example, the compiler infers nominal roles for foreign data types, which is safe but can be too constraining. For example, we should be able to coerce `Effect Age` to `Effect Int` if `Age` is a newtype over `Int`, but because `Effect` is a foreign data type its type parameter is inferred with a nominal role. We can loosen the role with an annotation:
+
+```purs
+foreign import data Effect :: Type -> Type
+
+-- this is a role annotation, setting Effect's type parameter to be
+-- representational instead of nominal (its inferred role).
+type role Effect representational
+```
+
+In contrast, it is unsafe to allow coercions between `Map k1 a` and `Map k2 a`, even when `Coercible k1 k2` holds. We can annotate `Map`'s key type parameter with a nominal role to prevent coercions:
```

Comment:
Two things. First, perhaps we should explain why this is unsafe? They can have two different `Ord` instances.

Second, it might help to clarify that it still is safe to coerce the values part of the Map.

Reply:
I'm finding it difficult to strike a balance in this section between a) here's just what you need to do to your code and b) here's what the new feature is all about. I don't want to give a tutorial, but as a new feature it does feel like there's quite a bit of requisite knowledge.

I think I ought to add that unless you know you want polymorphic kinds you should probably add a kind signature fixing everything to `Type`, and if you want something more sophisticated you should consult the release notes and the paper to write the annotation.

Reply:
> unless you know you want polymorphic kinds you should probably add a kind signature fixing everything to Type, and if you want something more sophisticated you should consult the release notes and the paper to write the annotation.

That sounds sensible.

Reply:
I'd prefer to write as little as possible here and link to more in-depth documentation like #371. @kl0tl, do you have any suggested phrasing for this section that would help folks get right to any code changes they may need to make and otherwise redirect them to better documentation?

Reply:
I actually think most - maybe even _all_ - of the discussion of roles could be struck from this document. My impression is that most users shouldn't need to care about roles, since role inference should cover the vast majority of cases. Also, since they're new, I think they're unlikely to break any existing code: it's only if you want to _start_ using Coercible that you might need to care, but this document is really aimed at people wanting to get their existing code compiling on 0.14.0, right?

Reply:
Just a thought, but would linking to [my 5-minute answer to a question on the matter](https://discourse.purescript.org/t/what-is-a-role/2109/2) with a comment of "If you're curious, see this link" suffice? I'll likely document this feature more in my learning repo once `v0.14.0` comes out.

Reply:
I don't think we need more places to read about roles. There's what will hopefully be the formal documentation in #371; there's the documentation for `Coercible` in purescript/purescript#3999, there's the changelog, there's this migration guide... if we do need to keep reexplaining roles for different audiences in this way, then fine, but let's at least give the references a star topology in terms of where they link to for more information. (Meaning, I think we should link to the document in #371 and no other document.)

Reply:
Fair point!

Reply:
I've updated this to just link to the document in #371 (anticipating its new location).